### ISSUE TITLE
Take screenshots of the browser for home tiles

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/firefox-tv#521]
Break-out from #442 because it's parallelizable. Here's what we know https://github.com/mozilla-mobile/firefox-tv/issues/442#issuecomment-367524464:

Screenshots are a bit messy and I haven't gotten any of these approaches to work with AmazonWebView in Firefox TV yet and all the testing I've done below is on emulators:
- You can get the drawn content of a View with `View.setDrawingCacheEnabled(true); val bitmap = getDrawingCache();`. This will only captured the visible area of the view (i.e. not the top of the page if the user has scrolled). This appears to work on AmazonWebView and WebView in a raw app
- You can call `view.draw(canvas)` where canvas wraps a bitmap you own. This is the approach you'd want to take to capture the whole screen (in combination with `enableSlowWholeDocumentDraw()`). This does not work for AmazonWebView in a raw app (bitmap is empty) but it works for WebView.
- There's a deprecated webView API: `capturePicture`. This isn't available on AmazonWebView.

This creates problems:
- We can't get a screenshot of the whole page with AmazonWebView. If we want a screenshot of the top of the page, we'll need to capture a screenshot before the user scrolls
- If we want an ideal screenshot, we also need to wait for the page to load before taking a screenshot.
- In order to capture the top of the page, we'll need to 1) screenshot every page we go to, regardless of whether the user pinned it (compromising on loaded, scrolled, and perf), 2) scroll the page to the top after the user presses pin (potentially compromising on UX), or 3) create a WebView in the background, wait for that to load, and screenshot that (assuming it's possible), compromising on data usage and performance. One upside to this approach: we can use the WebView, which seems to behave more nicely to screenshots.

Sounds like we may need to compromise between scrolled or not fully loaded.

---

Susheel mentions we could just capture the visible area for now and do the hard stuff later: https://github.com/mozilla-mobile/firefox-tv/issues/442#issuecomment-367560491

---

Note: I'll be smacking these screenshots together with the favicons in #442 (or another follow-up bug).

**Comment by mcomella** [comment]
Tif, here's a screenshot of the home tiles where I take the image and scale it down, without distortion, to fit in a home tile:

![screenshot_1519437630](https://user-images.githubusercontent.com/759372/36624418-be557d62-18c3-11e8-95af-cb08043c17f4.png)

What should the designs like for screenshots? i.e. how should I format the screenshot to 1) be of best value to users and 2) look the best? e.g. should I crop it to the top-left (like desktop does)? Should I just crop it such that it will fit in the box?

Keep in mind that because this home screen is an interim step, it doesn't need to be perfect. :) I discovered I can probably save the screenshots without any distortion so we can format it however we want now and later.

fwiw, here's the original screenshot:
![screenshot_1519437778](https://user-images.githubusercontent.com/759372/36624432-130b22ee-18c4-11e8-86f5-a82a1fafb93d.png)

**Comment by Sdaswani** [comment]
looks good to me!

**Comment by lime124** [comment]
@mcomella  i agree with susheel - i think that looks fine. it's hard to tell if the white border is part of the website or part of the tile - which is it?

can we keep that overlay that was on top of the tile to make them darker? or is that interferring with the picture too much? if so, then we should darken up the tile color as it looks like it's way too white.

**Comment by mcomella** [comment]
> it's hard to tell if the white border is part of the website or part of the tile - which is it?

The white border is the tile – you can click on the direct website screenshot above to see the raw image.

> can we keep that overlay that was on top of the tile to make them darker? or is that interfering with the picture too much?

I'll look into it.

**Comment by lime124** [comment]
i think i forgot to remove my assignment from the last question so i'm going to do that now. don't hesitate to ping me if i missed something!

**Comment by mcomella** [comment]
@lime124 What should we show on a tile if we can't get a screenshot for whatever reason? We could quickly borrow the Focus code (purple tile, domain letter) or the fennec code (colored tile, domain letter) if those designs work for you or we could try something else!

**Comment by mcomella** [comment]
@lime124 Also, the screenshots are being fetched asynchronously from disk so the images pop-in. To get around this, I added an alpha animation: https://drive.google.com/file/d/1vuhmAFtknztm9Zbk4CKCIEE5jqwqRrZx/view?usp=sharing

However, it looks funky. I suspect it's because:
- If some tile images are fading in, all of them should fade in
- If the images are fading in, we should fade the tiles too

How should we animate in these screenshots, Tif? Feel free to throw in more crazy ideas, e.g. we could also also fade in the whole home tiles section only after all of the assets are loaded.

**Comment by mcomella** [comment]
> can we keep that overlay that was on top of the tile to make them darker?

It turns out the overlay is not visible in the emulator (where I took the screenshot from) but is visible on device. Here's what we currently have:

![device-2018-03-01-175608](https://user-images.githubusercontent.com/759372/36879694-df31c6e8-1d79-11e8-991e-fe9cb7d3b900.png)

On Slack, Tif suggested lowering the tile background and removing the overlay. Here's what that could look like with an #f2f2f2 background:

![device-2018-03-01-180313](https://user-images.githubusercontent.com/759372/36879915-d7371c4e-1d7a-11e8-8349-2579c568755c.png)

Note that we still have the problem where the Youtube/Video Search assets have a white background.

@lime124 What do you think? How should I change the appearance of the screenshots?

**Comment by mcomella** [comment]
Also, here's a screenshot where I enlarged the screenshot size:

![device-2018-03-01-180643](https://user-images.githubusercontent.com/759372/36880009-5f53a12e-1d7b-11e8-84df-03ba4cc8de24.png)

I think we should probably increase their size, even if we don't increase the size of the existing icons. Maybe we could round their corners too. :P

**Comment by lime124** [comment]
@mcomella catching up...

re: animation ideas for screenshots popping in - let's catch up monday with epang about some ideas.

re: what do we so if we can't do a screenshot - if memory serves we've punted on favicons, even if they are fuzzy? how likely is this to happen? i would say that until we can do favicons, the focus way would be fine, but let's change the purple to something else. maybe ink?

re: title color being f2 - ugh i totally forgot that that's why we did it like that. let's leave the background color the it the way it is now. 

re: screenshot sizes - i like the enlarged version too and let's definitely round those corners.

Thanks!!!!
